local_archives/nwp/_generic_definitions_factory.py

```diff
@@ -0,0 +1,277 @@
+            # to see if it is an empty path.
+            fi, src = opts.fetcher.downloadToTemp(fi=fi)
+            if src is None or src == pathlib.Path():
+                raise ValueError(
+                    f"Error downloading file {fi.filename()}. See stdout logs for details."
+                )
+            context.log.info(
+                f"Moving file {src.as_posix()} to {dst.as_posix()}"
+            )
+            dst.parent.mkdir(parents=True, exist_ok=True)
+            shutil.move(src=src, dst=dst)
+
+            stored_paths.append(dst)
+            sizes.append(dst.stat().st_size)
+
+        elapsed_time = dt.datetime.now(tz=dt.UTC) - execution_start
+
+        yield dg.Output(
+            stored_paths,
+            metadata={
+                "inittime": dg.MetadataValue.text(context.asset_partition_key_for_output()),
+                "partition_num_files": dg.MetadataValue.int(len(stored_paths)),
+                "file_paths": dg.MetadataValue.text(str([f.as_posix() for f in stored_paths])),
+                "partition_size": dg.MetadataValue.int(sum(sizes)),
+                "area": dg.MetadataValue.text(opts.area),
+                "elapsed_time_mins": dg.MetadataValue.float(elapsed_time / dt.timedelta(minutes=1)),
+            },
+        )
+
+        # Perform the checks defined in the check_specs above
+        yield dg.AssetCheckResult(
+            check_name="num_local_is_num_remote",
+            passed=bool(len(stored_paths) == len(fis)),
+        )
+        yield dg.AssetCheckResult(
+            check_name="nonzero_local_size",
+            passed=bool(np.all(sizes)),
+        )
+
+    @dg.asset(
+        name="zarr_archive",
+        key_prefix=opts.key_prefix(),
+        partitions_def=opts.partitions,
+        ins={"raw_paths": dg.AssetIn(key=_raw_archive.key)},
+        io_manager_key="nwp_xr_zarr_io",
+        compute_kind="process",
+        metadata={
+            "archive_folder": dg.MetadataValue.text(f"{ZARR_FOLDER}/{'/'.join(opts.key_prefix())}"),
+            "area": dg.MetadataValue.text(opts.area),
+            "source": dg.MetadataValue.text(opts.source),
+        },
+        op_tags={"MAX_RUNTIME_SECONDS_TAG": 1000},  # TODO: Make more reasonable
+    )
+    def _zarr_archive(
+        context: dg.AssetExecutionContext,
+        raw_paths: list[pathlib.Path],
+    ) -> dg.Output[xr.Dataset]:
+        """Locally stored archive of zarr-formatted xarray data."""
+        execution_start = dt.datetime.now(tz=dt.UTC)
+        # Convert each file to an xarray dataset and merge
+        datasets: list[xr.Dataset] = []
+        for path in raw_paths:
+            context.log.info(
+                f"Converting raw file at {path.as_posix()} to xarray dataset."
+            )
+            datasets.append(opts.fetcher.mapTemp(p=path))
+        context.log.info(
+            f"Merging {len(datasets)} datasets into one."
+        )
+        ds = xr.merge(datasets, combine_attrs="drop_conflicts")
```

Comment:
This seems like it might be redoing what is in `service` in the NWP consumer, why don't we use the `service` here? It might save duplication, and things like the multi-level and special merging that is needed for GFS for example

Reply:
It's similar, although the service's "convert to zarr" step doesn't get the luxury of having all the raw paths to convert explicitly passed to it, like the dagster implementation does. So it would make slightly less use of the benefits of dagster, but it definitely could be done, and perhaps should!

Reply:
Also, the dagster implementation doesn't actually do any of the saving logic - that's handled by an IO manager so you can just return an xarray dataset which would make things easier downstream if we had any further asset dependencies, which we also couldn't do if we used service...

Reply:
Ah okay, in which case we probably should port over the merging logic I added to service then, so that it can properly merge datasets with different numbers of data variables and shapes.